File: reference/rocketride-server/packages/server/engine-lib/engLib/store/core/memory.cpp
// =============================================================================
// MIT License
// Copyright (c) 2026 Aparavi Software AG
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
// =============================================================================

#ifdef CLASSIFY_DLL_EXPORT
// When building as part of classify wrapper DLL, use apLib directly (no
// engLib/Python deps)
#include <apLib/ap.h>
#include <cstdlib>  // For malloc/free/realloc
#include <cstring>  // For memset

// Setup namespace aliasing like eng.h does - required for engLib headers
namespace engine {
using namespace ap;
}
namespace engine::store {
using namespace ap;
}

#include <engLib/store/headers/memory.hpp>
#else
// Normal engLib build uses the master header
#include <engLib/eng.h>
#endif

namespace engine::store {
//-------------------------------------------------------------------------
/// @details
///		Allocate memory
///	@param[in]	size
///		Size of block to allocate
///	@param[out]	ppBlockIn
///		Receives a ptr to the block
///	@returns
///		Error
//-------------------------------------------------------------------------
Error Memory::alloc(size_t size, void *ppBlockIn) {
    void **ppBlock = (void **)ppBlockIn;
    void *pBlock;

    // Allocate the block
    pBlock = ::malloc(size);

    // If out of memory, say so
    if (!pBlock)
        return APERR(Ec::OutOfMemory, _location,
                     "Out of memory allocating {} bytes", size);

    // Clear it
    memset(pBlock, 0, size);

    // Return the block
    *ppBlock = pBlock;
    return {};
}

//-------------------------------------------------------------------------
/// @details
///		Release memory
///	@param[in,out]	ppBlockIn
///		Releases a block or memory. Note it is a ptr to a ptr to the block of
///		memory. If the content of the ptr is null, there is no block. On return,
///		the ptr will be reset
///	@returns
///		Error
//-------------------------------------------------------------------------
Error Memory::release(void *ppBlockIn) {
    // Free it
    void **ppBlock = (void **)ppBlockIn;

    // If it is not allocated
    if (!*ppBlock) return {};

    // Release it
    ::free(*ppBlock);

    // Set the ptr to null
    *ppBlock = nullptr;
    return {};
}

//-------------------------------------------------------------------------
/// @details
///		Resize a block of memory
///	@param[in]	size
///		Size of block to allocate
///	@param[in,out]	ppBlockIn
///		Block of memory to resize. Note it is a ptr to a ptr to the block of
///		memory. If the content of the ptr is null, there is no block. On return,
///		the ptr will be reset
///	@returns
///		Error
//-------------------------------------------------------------------------
Error Memory::resize(size_t size, void *ppBlockIn) {
    void **ppBlock = (void **)ppBlockIn;
    void *pBlock;

    // Reallocate the block
    pBlock = ::realloc(*ppBlock, size);

    // If out of memory, say so
    if (!pBlock)
        return APERR(Ec::OutOfMemory, _location,
                     "Out of memory allocating {} bytes", size);

    // Return the block
    *ppBlock = pBlock;
    return {};
}
}  // namespace engine::store

File: reference/rocketride-server/packages/ai/src/ai/account/store_providers/memory.py
# MIT License
#
# Copyright (c) 2026 Aparavi Software AG
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

"""In-memory IStore implementation for testing."""

import re
from typing import Optional
import posixpath

from ..store import IStore, StorageError, VersionMismatchError


class MemoryStore(IStore):
    """
    In-memory storage backend.

    Stores all data in plain dicts. Versions are monotonic integers
    incremented on every write. Intended for unit and integration tests;
    data is not persisted across instances.
    """

    def __init__(self) -> None:
        super().__init__('memory://')
        self._files: dict[str, str] = {}
        self._versions: dict[str, int] = {}

    async def write_file(self, filename: str, data: str) -> None:
        filename = _check_path(filename)
        self._files[filename] = data
        self._versions[filename] = self._versions.get(filename, 0) + 1

    async def read_file(self, filename: str) -> str:
        filename = _check_path(filename)
        if filename not in self._files:
            raise StorageError(f'File not found: {filename}')
        return self._files[filename]

    async def read_file_with_metadata(self, filename: str) -> tuple:
        filename = _check_path(filename)
        if filename not in self._files:
            raise StorageError(f'File not found: {filename}')
        return self._files[filename], str(self._versions[filename])

    async def write_file_atomic(self, filename: str, data: str, expected_version: Optional[str] = None) -> str:
        filename = _check_path(filename)
        if expected_version is not None and filename in self._files:
            current = str(self._versions[filename])
            if current != expected_version:
                raise VersionMismatchError(
                    filename=filename,
                    expected_version=expected_version,
                    actual_version=current,
                )
        self._files[filename] = data
        self._versions[filename] = self._versions.get(filename, 0) + 1
        return str(self._versions[filename])

    async def delete_file(self, filename: str, expected_version: Optional[str] = None) -> None:
        filename = _check_path(filename)
        if filename not in self._files:
            raise StorageError(f'File not found: {filename}')
        if expected_version is not None:
            current = str(self._versions[filename])
            if current != expected_version:
                raise VersionMismatchError(
                    filename=filename,
                    expected_version=expected_version,
                    actual_version=current,
                )
        del self._files[filename]
        del self._versions[filename]

    async def list_files(self, prefix: str = '') -> list:
        prefix = _check_path(prefix)
        return sorted(f for f in self._files if f.startswith(prefix))

    async def list_entries(
        self,
        prefix: str = '',
        *,
        recursive: bool = True,
        include_files: bool = True,
        include_dirs: bool = True,
        name_pattern: Optional[str] = None,
    ) -> list:
        prefix = _check_path(prefix)

        # Basic check is enough for in-memory store for testing
        if name_pattern and ('/' in name_pattern or '\\' in name_pattern):
            raise StorageError(f'Invalid name pattern: {name_pattern}')
        if name_pattern == '..':
            raise StorageError(f'Path traversal detected: {name_pattern}')

        prefix_dir = prefix.rstrip('/')

        re_pattern = r'^' + ((re.escape(prefix_dir) + r'/') if prefix_dir else r'')
        if recursive:
            re_pattern += r'([^/]+/)*'
        if name_pattern:
            for c in name_pattern:
                if c == '*':
                    re_pattern += r'[^/]*'
                elif c == '?':
                    re_pattern += r'[^/]'
                else:
                    re_pattern += re.escape(c)
        else:
            re_pattern += r'[^/]+'
        re_pattern += r'$'
        re_pattern = re.compile(re_pattern)

        result = []
        seen_dirs = set() if include_dirs else None
        for filepath in self._files:
            if include_dirs:
                parts = filepath.split('/')[:-1]
                dir_path = ''
                for part in parts:
                    dir_path += ('/' if dir_path else '') + part
                    if dir_path not in seen_dirs:
                        seen_dirs.add(dir_path)
                        if re_pattern.match(dir_path):
                            result.append(dir_path + '/')

            if include_files and re_pattern.match(filepath):
                result.append(filepath)

        return sorted(result)


def _check_path(filename: str) -> str:
    # Basic check is enough for in-memory store for testing
    normalized = posixpath.normpath(filename.replace('\\', '/').strip('/'))
    if normalized.startswith('../') or normalized == '..':
        raise StorageError(f'Path traversal detected: {filename}')
    if normalized == '.':
        normalized = ''
    return normalized
